pinot-controller/src/main/java/org/apache/pinot/controller/workload/QueryWorkloadManager.java

```diff
@@ -170,34 +312,182 @@ public Map<String, InstanceCost> getWorkloadToInstanceCostFor(String instanceNam
         return workloadToInstanceCostMap;
       }
 
-      // Find all workloads associated with the helix tags
+      // Find all helix tags for this instance
+      InstanceConfig instanceConfig = _pinotHelixResourceManager.getHelixInstanceConfig(instanceName);
+      if (instanceConfig == null) {
+        LOGGER.warn("Instance config not found for instance: {}", instanceName);
+        return workloadToInstanceCostMap;
+      }
+
+      List<String> instanceTags = instanceConfig.getTags();
+      if (instanceTags == null || instanceTags.isEmpty()) {
+        LOGGER.warn("No tags found for instance: {}, cannot compute workload costs", instanceName);
+        return workloadToInstanceCostMap;
+      }
+
+      // Filter workloads by the instance's tags
       Set<QueryWorkloadConfig> queryWorkloadConfigsForTags =
-          PropagationUtils.getQueryWorkloadConfigsForTags(_pinotHelixResourceManager, instanceConfig.getTags(),
-                  queryWorkloadConfigs);
-      // Calculate the instance cost from each workload
+          PropagationUtils.getQueryWorkloadConfigsForTags(_pinotHelixResourceManager, instanceTags,
+              queryWorkloadConfigs);
+
+      if (queryWorkloadConfigsForTags.isEmpty()) {
+        LOGGER.debug("No workload configs match instance: {}", instanceName);
+        return workloadToInstanceCostMap;
+      }
+
+      // For each workload, aggregate contributions across all applicable nodeConfigs and propagation entities
       for (QueryWorkloadConfig queryWorkloadConfig : queryWorkloadConfigsForTags) {
+        String queryWorkloadName = queryWorkloadConfig.getQueryWorkloadName();
         for (NodeConfig nodeConfig : queryWorkloadConfig.getNodeConfigs()) {
-          if (nodeConfig.getNodeType() == nodeType) {
-            Set<String> instances = resolveInstances(nodeConfig);
-            InstanceCost instanceCost = _costSplitter.computeInstanceCost(nodeConfig, instances, instanceName);
-            if (instanceCost != null) {
-              workloadToInstanceCostMap.put(queryWorkloadConfig.getQueryWorkloadName(), instanceCost);
+          try {
+            if (nodeConfig.getNodeType() == nodeType) {
+              List<String> errors = QueryWorkloadConfigUtils.validateQueryWorkloadConfig(queryWorkloadConfig);
```

Comment:
Same here. Why are we validating workloadConfigs here? It should happen only during add/update of workload configs. Validating workload configs here and failing is too late.